src/Symfony/Component/Form/ChoiceList/ChoiceList.php

```diff
@@ -0,0 +1,101 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\ChoiceList;
+
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceFilter;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceFieldName;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceLabel;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceLoader;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\ChoiceValue;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\GroupBy;
+use Symfony\Component\Form\ChoiceList\Factory\Cache\PreferredChoice;
+use Symfony\Component\Form\ChoiceList\Loader\CallbackChoiceLoader;
+use Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface;
+use Symfony\Component\Form\FormTypeInterface;
+
+/**
+ * A set of convenient static methods to create choice list callbacks or loaders.
+ *
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+class ChoiceList
+{
+    /**
+     * @param FormTypeInterface     $formType A form type configuring a list
+     * @param ChoiceLoaderInterface $loader   Any pseudo callable to create a unique value from a choice
+     * @param mixed                 $vary     Dynamic data used to compute a unique hash when caching the loader
+     */
+    public static function lazy(FormTypeInterface $formType, callable $choices, $vary = null): ChoiceLoader
+    {
+        return self::loader($formType, new CallbackChoiceLoader($choices), $vary);
+    }
+
+    /**
+     * @param FormTypeInterface     $formType A form type configuring a list
+     * @param ChoiceLoaderInterface $loader   Any pseudo callable to create a unique value from a choice
+     * @param mixed                 $vary     Dynamic data used to compute a unique hash when caching the loader
+     */
+    public static function loader(FormTypeInterface $formType, ChoiceLoaderInterface $loader, $vary = null): ChoiceLoader
+    {
+        return new ChoiceLoader($formType, $loader, $vary);
+    }
+
+    /**
+     * @param FormTypeInterface $formType A form type configuring a list
+     * @param callable          $value    Any pseudo callable to create a unique value from a choice
+     * @param mixed             $vary     Dynamic data used to compute a unique hash when caching the callback
+     */
+    public static function value(FormTypeInterface $formType, $value, $vary = null): ChoiceValue
+    {
+        return new ChoiceValue($formType, $value);
+    }
+
+    /**
+     * @param FormTypeInterface $formType A form type configuring a list
+     * @param callable          $label    Any pseudo callable to create a label from a choice
+     * @param mixed             $vary     Dynamic data used to compute a unique hash when caching the callback
+     */
+    public static function label(FormTypeInterface $formType, $label, $vary = null): ChoiceLabel
```

Comment:
If $label is a callable (phpdoc) can you add a type hint here like above?

Reply:
Same for the following in the other methods

Reply:
No we can't, a string may be decorated to be a property path, hence "pseudo" callable description

Reply:
Thanks for the explanation this makes sense